src/ketu/async/source.clj

```diff
@@ -112,6 +113,10 @@
         ->data (->data-fn opts)
         put! (fn [record] (put-or-abort-pending! out-chan (->data record) abort-pending-put))
 
+        maybe-execute-custom-command (if commands-chan
+                                       (fn [] (when-let [command (async/poll! commands-chan)]
+                                                (command {:ketu.source/consumer consumer})))
+                                       (fn []))
```

Comment:
I wouldn't want an empty function call for the users who won't supply the `:ketu.source/consumer-commands-chan` option.  
Can you think of another implementation to avoid these calls?  

Maybe something like:
```clojure
(let [;...
        execute-custom-command! #(when-let [command (async/poll! commands-chan)]
                                   (command {:ketu.source/consumer consumer}))
        poll-and-put! #(let [records (poll!)]
                         (run! put! records))]

  ;...

  (if (some? commands-chan)
    (while [@should-poll?]
      (execute-custom-command!)
      (poll-and-put!))
    (while [@should-poll?]
      (poll-and-put!)))

;...
)
```

Reply:
I understand the concern about having a no-op function in each iteration of the polling loop, but from benchmarks that I have ran, it seems negligible (see below).  
I understand the suggestion about creating two "versions" of the polling loop (one with support for the custom command if it's available and one without) but I believe it would create problems in the future where each change in that area would need to be duplicated in both versions.

Here is the  benchmark I have ran (using criterium):
``` clojure
(let [poll-and-put! (fn [] (dotimes [i 1000] (* i i)))
      noop1 (fn [])
      noop2 (constantly nil)]
  (do (println "\n*** baseline ***")
      (criterium.core/bench (do (poll-and-put!))))
  (do (println "\n*** noop - empty fn ***")
      (criterium.core/bench (do (noop1) (poll-and-put!))))
  (do (println "\n*** noop - constantly ***")
      (criterium.core/bench (do (noop2) (poll-and-put!)))))
```

And here are the results:  
```
*** baseline ***
Evaluation count : 8077740 in 60 samples of 134629 calls.
             Execution time mean : 7.635551 µs
    Execution time std-deviation : 573.684180 ns
   Execution time lower quantile : 7.337461 µs ( 2.5%)
   Execution time upper quantile : 9.267974 µs (97.5%)
                   Overhead used : 6.673738 ns

Found 7 outliers in 60 samples (11.6667 %)
	low-severe	 1 (1.6667 %)
	low-mild	 6 (10.0000 %)
 Variance from outliers : 56.7636 % Variance is severely inflated by outliers

*** noop - empty fn ***
Evaluation count : 8096880 in 60 samples of 134948 calls.
             Execution time mean : 7.568388 µs
    Execution time std-deviation : 166.672843 ns
   Execution time lower quantile : 7.356883 µs ( 2.5%)
   Execution time upper quantile : 7.941921 µs (97.5%)
                   Overhead used : 6.673738 ns

Found 2 outliers in 60 samples (3.3333 %)
	low-severe	 2 (3.3333 %)
 Variance from outliers : 9.4630 % Variance is slightly inflated by outliers

*** noop - constantly ***
Evaluation count : 8157180 in 60 samples of 135953 calls.
             Execution time mean : 7.536907 µs
    Execution time std-deviation : 577.272610 ns
   Execution time lower quantile : 7.348291 µs ( 2.5%)
   Execution time upper quantile : 7.703311 µs (97.5%)
                   Overhead used : 6.673738 ns

Found 1 outliers in 60 samples (1.6667 %)
	low-severe	 1 (1.6667 %)
 Variance from outliers : 56.8228 % Variance is severely inflated by outliers
```